### DISCUSSION TITLE
Grafana v9.4.3 Table Panel Issue: Unable to perform operation on date column field

### DISCUSSION THREAD

**Discussion by lakshay2711** [grafana/grafana#78623] [General]
Hi,

I am rendering a table panel view with raw data metric using elastic-search data source and it successfully display the table. 
The requirement is to perform some operation on one column field of the table (returnTime) which is having a date format something like "2023-11-20 10:00:00" and I need a new column like "Delay" which basically does a simple math 

`toDateTime(returnTime) - now()`

To achieve this, I was using **Transform -> Add field from calculation -> Binary Operation** feature of grafana and it gives me two dropdowns to select the field or number. But I guess I am unable to use "toDateTime" and "now" functions in the dropdowns.

Can anyone please help me on this. Any help would be highly appreciated.

Thanks
Lakshay

**Comment by ismaildasci** [comment]
Hey Lakshay,

The Binary Operation transform in Grafana does not support functions like `toDateTime()` or `now()` directly in the dropdowns - it only works with existing numeric fields.

To calculate the delay between `returnTime` and current time, you have a few options:

**Option 1: Use the "Add field from calculation" transform with "Reduce row" mode**

If your date field is already parsed as a time field (epoch milliseconds), you can:
1. Add a transform: **Add field from calculation**
2. Mode: **Binary operation**
3. Operation: **Subtract (-)**
4. Select your `returnTime` field and subtract it from the current time

But since you need `now()`, this wont work directly.

**Option 2: Calculate in Elasticsearch query itself**

The cleanest approach is to do this calculation in your Elasticsearch query using a scripted field:

```json
{
  "script_fields": {
    "delay": {
      "script": {
        "source": "Instant.now().toEpochMilli() - doc['returnTime'].value.toInstant().toEpochMilli()",
        "lang": "painless"
      }
    }
  }
}
```

This returns delay in milliseconds. You can then format it in Grafana using field overrides.

**Option 3: Use "Add field from calculation" with "Row index" + clever math**

If you export the dashboard JSON and set the `now` time as a calculation baseline using expressions, but this gets complicated.

**Recommended approach:**

Do the calculation server-side in Elasticsearch. Its more reliable and the transform layer in Grafana is really meant for simpler field manipulations rather than time-based calculations involving `now()`.

Hope this helps!
